src/librustc/traits/error_reporting.rs

```diff
@@ -544,68 +597,84 @@ pub fn report_selection_error<'a, 'tcx>(infcx: &InferCtxt<'a, 'tcx>,
+        Some(_) => None,
+        None => {
+            Some(struct_span_err!(
+                tcx.sess, span, E0038,
+                "the trait `{}` cannot be made into an object",
+                tcx.item_path_str(trait_def_id)))
+        }
+    };
 
     let mut reported_violations = FnvHashSet();
     for violation in violations {
         if !reported_violations.insert(violation.clone()) {
             continue;
         }
-        match violation {
+        let buf;
+        let note = match violation {
             ObjectSafetyViolation::SizedSelf => {
-                err.fileline_note(
-                    span,
-                    "the trait cannot require that `Self : Sized`");
+                "the trait cannot require that `Self : Sized`"
             }
 
             ObjectSafetyViolation::SupertraitSelf => {
-                err.fileline_note(
-                    span,
-                    "the trait cannot use `Self` as a type parameter \
-                     in the supertrait listing");
+                "the trait cannot use `Self` as a type parameter \
+                     in the supertrait listing"
             }
 
             ObjectSafetyViolation::Method(method,
                                           MethodViolationCode::StaticMethod) => {
-                err.fileline_note(
-                    span,
-                    &format!("method `{}` has no receiver",
-                             method.name));
+                buf = format!("method `{}` has no receiver",
+                              method.name);
+                &buf
             }
 
             ObjectSafetyViolation::Method(method,
                                           MethodViolationCode::ReferencesSelf) => {
-                err.fileline_note(
-                    span,
-                    &format!("method `{}` references the `Self` type \
+                buf = format!("method `{}` references the `Self` type \
                               in its arguments or return type",
-                             method.name));
+                              method.name);
+                &buf
             }
 
             ObjectSafetyViolation::Method(method,
                                           MethodViolationCode::Generic) => {
-                err.fileline_note(
+                buf = format!("method `{}` has generic type parameters",
+                              method.name);
+                &buf
+            }
+        };
+        match (warning_node_id, &mut err) {
+            (Some(node_id), &mut None) => {
+                tcx.sess.add_lint(
+                    ::lint::builtin::OBJECT_UNSAFE_FRAGMENT,
+                    node_id,
                     span,
-                    &format!("method `{}` has generic type parameters",
-                             method.name));
+                    note.to_string());
             }
+            (None, &mut Some(ref mut err)) => {
+                err.fileline_note(span, note);
```

Comment:
Now that https://github.com/rust-lang/rust/pull/32756 has landed, we'll want to change these to just calls to `err.note`